### ISSUE TITLE
[OAuth2] ErrorAccessDenied. Token has no role when using OAuth2 codeunit

### ISSUE THREAD

**Issue by JulienDecottignies** [microsoft/AL#6151]
Hi,

I've created an app on Azure that give permission to **https://graph.microsoft.com/Calendars.ReadWrite** and User.Read.All (as Application Type). 
To authenticate, I use **OAuth2.AcquireTokenByAuthorizationCode()** function, and I receive correctly the token I requested.

To use this token in a graph request, I use **OAuth2.AcquireAuthorizationCodeTokenFromCache()** function that gives me the exactly same token as OAuth2.AcquireTokenByAuthorizationCode() sent me, and I put it in the authorization header of my request as Bearer {token}.
When I make a request to graph api on the calendar I have an **ErrorAccessDenied** exception.

When I analyze this token using https://jwt.ms/ **I have no role**.

When I request for a token with postman with the same setup, the token I receive has the correct roles:
![image](https://user-images.githubusercontent.com/29598240/93488029-2f7c0b80-f906-11ea-9405-6ed73738a535.png)

Is there a problem with how I use this function ? Why does my token does not have any role when I use OAuth2 codeunit?

Here is my code sample:

```
    procedure GetAccessToken(var OutAzureAppInfos: Record "GRA Azure App Infos")
    var
        OAuth2: Codeunit OAuth2;
        Scopes: List of [Text];
        ClientSecret: Text;
        AuthUrlLbl: Label 'https://login.microsoftonline.com/%1/oauth2/v2.0/authorize', Comment = '%1: The tenant ID';
        AccessToken: Text;
        ErrorTok: Text;
    begin
        Scopes.add('https://graph.microsoft.com/.default'); // Add all scopes defined for graph
       ClientSecret := //Retrieving ClientSecret here

        OAuth2.AcquireTokenByAuthorizationCode(OutAzureAppInfos.ClientId, ClientSecret, StrSubStNo(AuthUrlLbl, OutAzureAppInfos.Tenant), OutAzureAppInfos.RedirectUrl, Scopes, enum::"Prompt Interaction"::None, AccessToken, ErrorTok);


...
    end;
```

Where:
OutAzureAppInfos.RedirectUrl = https://businesscentral.dynamics.com/OAuthLanding.htm
OutAzureAppInfos.Tenant = The tenant where my azure app is hosted 

Note that I've also tried to Add the scope in the list individually instead of using https://graph.microsoft.com/.default.

Thanks for any help

**Comment by qutreson** [comment]
This repository no longer handles general questions and is only dedicated to bug reports for the latest preview version of the AL compiler and developer tools.

To get help for your question please use one of the following resources:
[Business Central Community](https://community.dynamics.com/business/f/758)
[mibuso forum](https://forum.mibuso.com/categories/nav-three-tier)
[Dynamics User Group](https://dynamicsuser.net/nav/f/developers)

**Comment by JulienDecottignies** [comment]
Hi @qutreson ,

I don't know if it's related to the way I use this function or _if the function itself_ has a problem. 
The searches I made for the ErrorAccessDenied exception, states for some that the token used for the request is not correct (as there is a **token_id** and an **access_token** returned by the /token endpoint). Using the token_id instead of access_token can throw this exception (as it does not contain the roles). So I wonder if this function sends back a wrong token (for my use). 

If the purpose of this function is to send this token (without roles so probably the token_id) and not another, then I will use another method of my own to get  the token (access_token) containing my app roles.
If the purpose of this function is to send back an access token containing roles, there is the possibility that I misuse it **or there is an issue with the AccessToken out parameter.** This issue I created is to handle this second case.

I've cloned the ALAppExtension repository (I've created this issue in this repo before but it has been moved here) to see what the code does exactly, and so what AccessToken out parameter is containing. But the code is using a dll that I cannot see inside, but this is where my answers are:
![image](https://user-images.githubusercontent.com/29598240/93740367-a1946f00-fbea-11ea-9c0d-5c7239a207e5.png)


Thanks
